First level model

MartinPerez opened the pull request:
1. First level model object basic implementation ready. 

2. I added trivially modified examples newapi_localizer and newapi_fiac that will keep being modified as the roadmap evolves. To compare with real simple examples how things are changing. For now changes are minimal. The localizer example shows better that now the design matrix do not need to be precomputed, while the fiac example shows that previous behavior, passing directly design matrices, still works.

3. memory issues were handled by releasing unnecessary variables and allowing the user to get rid of some regression results that are not useful unless the model will be inspected afterwards in details. The parameter minimize_memory in FirstLevelModel object takes care of this.

Improvements over issue #70 can be appreciated

![localizer](https://cloud.githubusercontent.com/assets/8766915/15839839/27d4ee6a-2c46-11e6-8df4-9289ef939d1f.png)

![fiac](https://cloud.githubusercontent.com/assets/8766915/15839840/2abd836c-2c46-11e6-913b-f91256b89e1f.png)


4. parallelization on OLS fit by voxel groups and on AR model by assigning each AR model to a job

5. issue #44 addressed by adding scaling_axis parameter. renamed percent_signal_change to signal_scaling.

6. Factored out contrasts functionality to their own script, for future coming improvements like formulas handling.

Note: glm.py is supposed to be completely replaced by first_level_model.py and contrasts.py. Its kept such that previous examples keep working.

Please let me know what you think.

MartinPerez commented:
@bthirion tests are still missing. Besides reusing glm.py tests, do you think I should test any new features in more details. For example different voxel scalings added?

MartinPerez commented:
Note: Only first_level_model.py, regression.py and utils.py require to be reviewed. The rest is moving code around. And the examples are almost identical to the original versions for now.

bthirion commented on examples/newapi_fiac_analysis.py:
```diff
@@ -0,0 +1,88 @@
+"""
```
We should remove examples with the old API, or this will become a mess.

bthirion commented on examples/newapi_localizer_analysis.py:
```diff
@@ -0,0 +1,93 @@
+"""
```
Same comment here. The example should inly use the best API, so please remove plot_localizer_analysis.py and rename newapi_localizer_analysis.py to plot_localizer_analysis.py.

bthirion commented on nistats/first_level_model.py:
```diff
@@ -0,0 +1,606 @@
+
+It is important to note that the GLM is meant as a one-session General Linear
+Model. But inference can be performed on multiple sessions by computing fixed
+effects on contrasts
```
This point no longer holds.

bthirion commented on nistats/contrasts.py:
```diff
@@ -0,0 +1,241 @@
+import numpy as np
```
Please remove all the corresponding code from glm.py. I am quite anxious about code duplication.

bthirion commented on nistats/first_level_model.py:
```diff
@@ -0,0 +1,606 @@
+
+
+def mean_scaling(Y, axis=0):
+    """Scaling of the data to have percent of baseline change columnwise
```
not columnwise, along the specified axis.

bthirion commented on nistats/first_level_model.py:
```diff
@@ -0,0 +1,606 @@
+    return results
+
+
+def _OLSModel_fit(X, Y):
```
_ols_model_fit

bthirion commented on nistats/first_level_model.py:
```diff
@@ -0,0 +1,606 @@
+
+
+def _minimize_memory_regression_results(results):
+    del results.Y
```
Please provide a one-line docstring

bthirion commented on nistats/first_level_model.py:
```diff
@@ -0,0 +1,606 @@
+    del results.model
+    del results.wY
+    del results.wresid
+    return results
```
I understand what youd o here, but I find it very dirty. Why not creating another simplified object ?

bthirion commented on nistats/first_level_model.py:
```diff
@@ -0,0 +1,606 @@
+    return OLSModel(X).fit(Y)
+
+
+def _ARModel_fit(X, val, Y):
```
_ar_model_fit

bthirion commented on nistats/first_level_model.py:
```diff
@@ -0,0 +1,606 @@
+                format(X.shape, Y.shape))
+
+    # Create the model
+    # model = OLSModel(X)
```
remove this line

bthirion commented on nistats/first_level_model.py:
```diff
@@ -0,0 +1,606 @@
+    # model = OLSModel(X)
+    # Parallelize model fit by voxel groups
+    n_voxels = Y.shape[1]
+    group_iter = GroupIterator(n_voxels, n_jobs)
```
gloups, I don't know that guy, and I don't like his name.

bthirion replied:
Is all this worth the added complexity for simple OLS fit ?

bthirion commented on nistats/first_level_model.py:
```diff
@@ -0,0 +1,606 @@
+
+
+def _check_list_length_match(list_1, list_2, var_name_1, var_name_2):
+    if len(list_1) != len(list_2):
```
This function should have a (1-line) docstring and it should not be here, but in utils.py

bthirion commented on nistats/first_level_model.py:
```diff
@@ -0,0 +1,606 @@
+            % (var_name_1, len(list_1), var_name_2, len(list_2)))
+
+
+def _check_and_load_tables(tables_, var_name):
```
Same here

bthirion commented on nistats/first_level_model.py:
```diff
@@ -0,0 +1,606 @@
+    return tables
+
+
+def _check_run_tables(run_imgs, tables_, tables_name):
```
idem

bthirion commented on nistats/first_level_model.py:
```diff
@@ -0,0 +1,606 @@
+            # Compute GLM
+            Y = self.masker_.transform(run_img)
+
+            # Y = self.masker_.transform(run_img)
```
Please remove these ghosts

bthirion commented on nistats/first_level_model.py:
```diff
@@ -0,0 +1,606 @@
+        else:
+            return self.design_matrices_
+
+    def _all_runs_contrast(self, con_vals, stat_type):
```
Please add a one-line docstring
Could this be a function instead of a method ?

bthirion commented on nistats/regression.py:
```diff
@@ -330,3 +329,17 @@ def SSE(self):
+                                axis=0)
+    cov = results[0].cov
+    return RegressionResults(theta, Y, model, wY, wresid,
+                             dispersion=dispersion, cov=cov)
```
I am a bit worried about np.concatenate, that can lead to unexpected behavior.
If all this is only used for OLS regression, I do not think that it is worth the effort.

bthirion commented:
There are some great API improvements ni this PR, thanks.
Could reduce code duplication whenever possible and test all the functions ? We cannot move forward without tests.

MartinPerez commented:
@bthirion thanks for the prompt feedback. I wanted your first feedback before eliminating all the other code. I will of course adapt the tests. I would agree with having a more simple regression results object and check afterwards if any extra complexity would be needed for other cases. I just wanted to show how the parallelization of OLS would look like, I will take it out and later test on big datasets the code to propose the change if its worth it.

MartinPerez commented:
@bthirion all comments addressed, all tests adapted and all examples adapted. This is ready for a second round of review

MartinPerez commented:
I added many more parameter checks and now compute contrast can take a single contrast definition for multiple runs. I think its safe to assume in that case we want the same contrast for all runs. Still is possible to provide one contrast per run to avoid including some of them.

bthirion commented on nistats/first_level_model.py:
```diff
@@ -0,0 +1,561 @@
+        This parameter is also passed to nilearn.signal.clean.
+        Please see the related documentation for details.
+
+    slice_time_ref: float, optional (default 0.)
```
The problem with your approach is that you no longer support irregular sampling, that was mad possible by directly providing frame_times. Leaving the possibility to use frame_times would yield now 3 arguments instead of 1. I understand that it makes the general case simpler (e.g. using BIDS convention that provide a TR and default slice_time, the frame_times would be inferred automatically).

bthirion commented on nistats/contrasts.py:
```diff
@@ -0,0 +1,261 @@
+
+
+def _summary_contrast(labels, results, con_vals, stat_type):
+    """Computes the summary contrast assuming fixed effects.
```
then it should be called _fixed_effect_contrast

bthirion commented on nistats/first_level_model.py:
```diff
@@ -0,0 +1,561 @@
+    return Y, mean
+
+
+def _ar_model_fit(X, val, Y):
```
I think we should avoid such one-liners. It makes debugging and reviewing harder and does not really help.

bthirion commented on nistats/first_level_model.py:
```diff
@@ -0,0 +1,561 @@
+
+    min_onset : float, optional
+        This parameter specifies the minimal onset relative to the design
+        (in seconds). Events that start before (slice_time_ref * t_r +
```
slice_time_ref + min_onset

bthirion commented on nistats/first_level_model.py:
```diff
@@ -0,0 +1,561 @@
+
+    subject_id : string, optional
+        id of the subject of this first_level_model instance. Will be used
+        to create subject specific memory cache if caching allowed.
```
I don't think we need this feature. This signatures are becoming horrible, so I'd like to avoid that parameter.

bthirion commented on nistats/first_level_model.py:
```diff
@@ -0,0 +1,561 @@
+        This parameter is passed to nilearn.signal.clean.
+        Please see the related documentation for details.
+
+    high_pass: False or float, optional
```
parameters clash: we now also have a period_cut parameter that does the same thing.
I think that the masker should be called with low_pass/high_pass parameters set to False automatically, so that all the filtering is explicit in the design matrix. the low_pass and high_pass parameters would then disappear from this class constructor.

bthirion commented on nistats/first_level_model.py:
```diff
@@ -0,0 +1,561 @@
+        Incompatible with standardize (standardize=False is enforced when\
+        signal_scaling is True).
+
+    scaling_axis: int or tuple, optional,
```
signal_scaling and scaling_axis should become one single parameter, that would be one of {False, 0, 1, (0,1)}

bthirion commented on nistats/first_level_model.py:
```diff
@@ -0,0 +1,561 @@
+        Returns
+        -------
+        params: dict
+            The dict of parameters
```
This docstring is not very informative (was already here before, maybe...).

bthirion replied:
Oops, actually this function is never used. Please remove it.

bthirion commented on nistats/first_level_model.py:
```diff
@@ -0,0 +1,561 @@
+        """ Fit the GLM
+
+        For each run:
+        1. create design matrix
```
1. create design matrix X

bthirion commented on nistats/first_level_model.py:
```diff
@@ -0,0 +1,561 @@
+                n_scans = run_img.get_data().shape[3]
+                if confounds is not None:
+                    confounds_matrix = confounds[run_idx].values
+                    if confounds_matrix.shape[0] != n_scans:
```
this has laready been checked in l.379. Please remove.

bthirion replied:
I undertsand that the check is different, but please add inline comments to help readers.

bthirion commented on nistats/first_level_model.py:
```diff
@@ -0,0 +1,561 @@
+                    confounds_names = None
+                start_time = self.slice_time_ref * self.t_r
+                end_time = (n_scans - self.slice_time_ref) * self.t_r
+                frame_times = np.linspace(start_time, end_time, n_scans)
```
Here does the assumption of regular acquisition comes in. Since the only person I know who does not use such this scheme is your supervisor, I let you decide whether this is important or not ;-)

bthirion commented on nistats/first_level_model.py:
```diff
@@ -0,0 +1,561 @@
+        if self.design_matrices_ is None:
+            raise ValueError('The model has not been fit yet')
+        else:
+            return self.design_matrices_
```
I think we should remove such getters that do exactly nothing.
writing model.design_matrices_ is not more complicated than model.get_design_matrices()

bthirion commented on nistats/first_level_model.py:
```diff
@@ -0,0 +1,561 @@
+            con_vals = con_vals * n_runs
+        if isinstance(output_type, _basestring):
+            if output_type not in ['z_score', 'stat', 'p_value', 'eff', 'var']:
+                raise ValueError('output_type must be "z_score", "stat",'
```
must be one of

bthirion commented on nistats/regression.py:
```diff
@@ -330,3 +329,65 @@ def SSE(self):
         return self.SSE / self.df_resid
+
+
+class SimpleRegressionResults(LikelihoodModelResults):
```
We'll probably have to refactor this later, but for the moment I'm fine with it as is.

bthirion commented on nistats/utils.py:
```diff
@@ -7,10 +7,44 @@
 py3 = sys.version_info[0] >= 3
 
 
+def _check_list_length_match(list_1, list_2, var_name_1, var_name_2):
```
Sorry, but all thee guys need to come with a test.

bthirion commented on nistats/first_level_model.py:
```diff
@@ -0,0 +1,561 @@
+        self.low_pass = low_pass
+        self.high_pass = high_pass
+        self.smoothing_fwhm = smoothing_fwhm
+        if memory is not None:
```
A test for that feature is welcome too.

bthirion commented on nistats/first_level_model.py:
```diff
@@ -0,0 +1,561 @@
+                verbose=max(0, self.verbose - 1),
+                target_shape=self.target_shape,
+                memory_level=self.memory_level)
+        else:
```
The following lines are not covered by any test

bthirion commented on nistats/first_level_model.py:
```diff
@@ -0,0 +1,561 @@
+
+            # Build the experimental design for the glm
+            run_img = check_niimg(run_img, ensure_ndim=4)
+            if design_matrices is None:
```
The following lines are not covered by tests.

bthirion commented:
Ok, this is looking better and better. We just need a few more tests and remove some functions that are not really necessary imho and make the API a bit heavy.

MartinPerez commented:
I added many more tests, particularly for errors raised and contrast and utility functions. Also addressed all comments as we discussed too. The only thing still lacking a bit is more tests after l. 389.

MartinPerez commented:
@bthirion I think I added enough tests for all first level model methods and raised errors. Please let me know what you think.

MartinPerez commented:
interesting that the ar model was not giving the exact same ar partitions for labels.

MartinPerez commented:
@bthirion i dont understand why in python 3.4 AR models is do not have consistent results. Should I open an issue to take care of this later? since its not directly relevant to the api reformulation.

bthirion commented:
Hm, the  failure must be quite subtle, because I don't see by eye where the two vectors are different. In any case, we need to fix this travis failure.

MartinPerez commented:
@bthirion I am confident the failure is really unrelated with this PR. I did not manage to solve it quickly, I opened an issue and commented the part of the test that I added that highlighted the problem. Please lets put that aside to merge this after passing review. The second level object and contrast formulas are already finished in the next PR too.

bthirion commented on nistats/contrasts.py:
```diff
@@ -0,0 +1,262 @@
+        self.tiny = tiny
+        self.dofmax = dofmax
+
+    def eff(self):
```
Do we need this ?

MartinPerez replied:
I use to simplify the code in compute_contrast, since after getting the contrast object I can call any method like 'stat' or 'p_value', including then 'eff' and 'var'.

bthirion replied:
then can them `effect()` and `variance()`.
If they are bound to be used only there, to avoid redundancy with the class parameters, I would have a unique `summary_statistics()` method that returns both.

MartinPerez replied:
what I mean is that in l. 547 of first_level_model.py I call the output_type as an attribute of the contrast object to simplify the code. A summary_statistics() method would make me revert to the switch logic that existed before I made this change. I thought this way was cleaner than handling each possible output_type with an if statement. I also thought of using the @setattr_on_read tag on the z_score, p_value and stat methods instead of creating the dummy functions. What do you prefer?

bthirion replied:
I see.
* for the moment, rename `eff` to `effect` at least or, even better, 'effect_size'. Similarly 'var' should be 'variance'.
* add an inline comment to explain the purpose of this function e.g. 'make access to summary statistics more straightforward when computing contrasts'
*  @setattr_on_read that makes sense. I have no strong feeling about it.

bthirion commented on nistats/contrasts.py:
```diff
@@ -0,0 +1,262 @@
+    def eff(self):
+        return self.effect[0, :]
+
+    def var(self):
```
Do we need this ?

MartinPerez replied:
same

bthirion commented on nistats/contrasts.py:
```diff
@@ -0,0 +1,262 @@
+        self.p_value_ = None
+        self.baseline = 0
+        self.tiny = tiny
+        self.dofmax = dofmax
```
I'm realizing the existence of this `dofmax` parameter. Apparently, it was already here before, but I'm quite worried that this is mostly meaningless and should probably be removed in the future.

bthirion replied:
Opened issue #79.

bthirion commented on nistats/first_level_model.py:
```diff
@@ -0,0 +1,515 @@
+    ----------
+
+    t_r: float
+        This parameter indicates repetition times of the experimental runs.
```
in seconds.

bthirion commented on nistats/first_level_model.py:
```diff
@@ -0,0 +1,515 @@
+
+    fir_delays : array of shape(n_onsets) or list, optional
+        In case of FIR design, yields the array of delays used in the FIR
+        model.
```
in seconds

bthirion commented on nistats/first_level_model.py:
```diff
@@ -0,0 +1,515 @@
+
+    period_cut : float, optional
+        This parameter specifies the cut period of the low-pass filter in
+        seconds for the design matrices.
```
in seconds

bthirion commented on nistats/tests/test_dmtx.py:
```diff
@@ -159,59 +157,29 @@ def test_design_matrix0d():
-    X, names = design_matrix_light(frame_times, paradigm, hrf_model=hrf_model,
-                            drift_model='polynomial', drift_order=3)
-    assert_equal(len(names), 7)
-
```
Why do you remove this test ?

MartinPerez replied:
all the matrix1* tests are now together in test_design_matrix1 in line 172. There were basically 4 tests duplicating all lines of code except the last one. I did not see the point of having 4 separate tests.

bthirion replied:
OK.

bthirion commented on nistats/tests/test_dmtx.py:
```diff
@@ -159,59 +157,29 @@ def test_design_matrix0d():
-    hrf_model = 'glover'
-    X, names = design_matrix_light(frame_times, paradigm, hrf_model=hrf_model,
-                        drift_model='polynomial', drift_order=3)
     assert_true((X[:, - 1] == 1).all())
```
Same question here.

MartinPerez replied:
same as above

bthirion commented on nistats/tests/test_first_level_model.py:
```diff
@@ -0,0 +1,308 @@
+                ValueError, FirstLevelModel(mask=None).fit, fi, d, conf)
+
+
+def basic_paradigm():
```
This test should not be here but in test_dmtx.py or test_experimental_paradigm.py

MartinPerez replied:
basic paradigm is an utility function used by the tests, not a test itself.

bthirion commented on nistats/tests/test_first_level_model.py:
```diff
@@ -0,0 +1,308 @@
+        # ar not giving consistent results in python 3.4
+        # assert_almost_equal(labels1, labels2, decimal=1) ####FIX
+        # assert_equal(len(results1), len(results2)) ####FIX
+
```
See issue #77

bthirion commented:
Aside from minor points, this is almost OK.

MartinPerez commented:
@bthirion I updated the documentation with seconds and answered your comments.

MartinPerez commented:
@bthirion updated the contrasts interface, compute contrast method and documentation and tests accordingly.

bthirion commented on nistats/first_level_model.py:
```diff
@@ -0,0 +1,516 @@
+                    confounds_matrix = None
+                    confounds_names = None
+                start_time = self.slice_time_ref * self.t_r
+                end_time = (n_scans - self.slice_time_ref) * self.t_r
```
Actually 
`end_time = (n_scans - 1 + self.slice_time_ref) * self.t_r`

bthirion commented:
Could you take care of this last fix (one line), then I'll merge ?

MartinPerez commented:
@bthirion done

bthirion commented:
All right, merging.
